rust/main/chains/hyperlane-starknet/src/validator_announce.rs

```diff
@@ -0,0 +1,225 @@
+        let storage_location = string_to_cairo_long_string(&announcement.value.storage_location)
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+        let signature_bytes = announcement.signature.to_vec();
+        let signature = &to_strk_message_bytes(&signature_bytes);
+
+        let tx = self
+            .contract
+            .announce(&EthAddress(validator), &storage_location, signature);
+
+        let gas_estimate = tx
+            .estimate_fee()
+            .await
+            .map_err(|e| HyperlaneStarknetError::AccountError(e.to_string()))?
+            .overall_fee;
+
+        let max_cost = gas_estimate * FieldElement::TWO;
+
+        Ok((tx.max_fee(max_cost), max_cost))
+    }
+
+    #[allow(unused)]
+    pub fn contract(
+        &self,
+    ) -> &StarknetValidatorAnnounceInternal<SingleOwnerAccount<AnyProvider, LocalWallet>> {
+        &self.contract
+    }
+}
+
+impl HyperlaneChain for StarknetValidatorAnnounce {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.provider.domain()
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl HyperlaneContract for StarknetValidatorAnnounce {
+    fn address(&self) -> H256 {
+        self.contract.address.into()
+    }
+}
+
+#[async_trait]
+impl ValidatorAnnounce for StarknetValidatorAnnounce {
+    async fn get_announced_storage_locations(
+        &self,
+        validators: &[H256],
+    ) -> ChainResult<Vec<Vec<String>>> {
+        let validators_calldata: Vec<EthAddress> = validators
+            .iter()
+            .map(|v| {
+                TryInto::<FieldElement>::try_into(*v).map_err(Into::<HyperlaneStarknetError>::into)
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .map(EthAddress)
+            .collect();
+
+        let storage_locations_res = self
+            .contract
+            .get_announced_storage_locations(&validators_calldata)
+            .call()
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+
+        let storage_locations = storage_locations_res
+            .into_iter()
+            .map(|validator_vec| {
+                validator_vec
+                    .into_iter()
+                    .map(|inner_vec| {
+                        inner_vec
+                            .into_iter()
+                            .map(|element| parse_cairo_short_string(&element))
+                            .collect::<Result<Vec<String>, ParseCairoShortStringError>>()
+                    })
+                    .collect::<Result<Vec<Vec<String>>, ParseCairoShortStringError>>()
+                    .map(|strings_vec| {
+                        strings_vec
+                            .into_iter()
+                            .map(|inner_vec| inner_vec.join(""))
+                            .collect::<Vec<String>>()
+                    })
+            })
```

Comment:
could you add a comment to clarify what's going on here? Not sure I understand

Reply:
added